### ISSUE TITLE
OHPL Reference and Conceptual Understanding

### ISSUE THREAD

**Issue by calvinwy** [ohpl/ohpl#1]
@sh-shmss 

hi Shayan, 

I was reading the paper on OHPL and I wanted to understand the algorithm better.  I noticed that there are 2 papers available at the moment and one of them is on the original OHPL-Net and there is another one is on Net Promoter Sentiment with OHPL-all.  The code in your repository appears to be OHPL-all, and in the Net Promoter Sentiment paper it says that there will be a new publication coming out to explain OHPL-all, but it doesn't seem to be available yet.  Or, do you have some kind of manuscript you can share?

I was reading your code and having a hard time understanding the implementation of the HPL section:

```
    # OHPL point loss
    # mean distance for each class
    yO = tf.transpose(y_pred) @ y_true_1hot
    yc = tf.reduce_sum(y_true_1hot,0)
    class_mean = tf.divide(yO,yc) 
 
    # mean by class
    class_mean = tf.reshape(class_mean,[-1,1])
    mean_matrix = y_true_1hot @ class_mean
    
    lower_bound = tf.subtract(min_label,y_true)
    lower_bound = tf.add(lower_bound,1)
    lower_bound = tf.multiply(lower_bound,1e9)
    # apply ReLU    
    lower_bound = tf.nn.relu(lower_bound)
    lower_bound = tf.add(margin, lower_bound)

    upper_bound = tf.subtract(y_true,max_label)
    upper_bound = tf.add(lower_bound,1)
    upper_bound = tf.multiply(lower_bound,1e9)
    # apply ReLU    
    upper_bound = tf.nn.relu(lower_bound)
    upper_bound = tf.add(margin, lower_bound)    

```

1. I don't quite understand why are we subtracting min_label from class labels?
2. Why are we adding 1 after the subtraction?
3. Why are we multiplying 1e9, is it some sort of hyperparameter?
4. Is this the same margin comparing to OHPL-net?
5. How to interpret the above code geometrically?

Do you mind to give me a hand explaining the concept on the above?  I am really interested in trying out the algorithm, but I just want to make sure that I understand what it is doing.  Many thanks!

Calvin

**Comment by calvinwy** [comment]
Okay, I think I got it.  The max_label and min_label are used to limit the unnecessary optimization of the edge cases.  However, how did you come up with the loss bound 1e9 for the edge cases?

**Comment by calvinwy** [comment]
One more point I am still a bit unsure about is the following:

```
# keeps min. distance
keep = tf.minimum(min_distance,1)
```

What is the purpose of the above?
